File: hobbies/unsorted/max-square/MaxSquare.java
import java.util.Scanner;

public class MaxSquare {
    public static void printMatrix(int[][] matrix) {
        for (int i = 0; i < matrix.length; ++i) {
            for (int j = 0; j < matrix[i].length; ++j) {
                System.out.printf("%d ", matrix[i][j]);
            }
            System.out.printf("%n");
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int row = sc.nextInt();
        int col = sc.nextInt();

        int[][] numbers = new int[row][col];
        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                numbers[i][j] = sc.nextInt();
            }
        }

        printMatrix(numbers);

        int[][] result = new int[row][col];

        System.arraycopy(numbers[0], 0, result[0], 0, row);

        for (int j = 0; j < col; ++j) {
            result[0][j] = numbers[0][j];
        }

        for (int i = 1; i < row; ++i) {
            for (int j = 1; j < col; ++j) {
                if (numbers[i][j] == 0) {
                    result[i][j] = 0;
                } else {
                    result[i][j] = Math.min(Math.min(result[i][j-1], result[i-1][j]),
                                            result[i-1][j-1]) + 1;
                }
            }
        }

        printMatrix(result);

        int max_s = result[0][0], max_i = 0, max_j = 0;
        for (int i = 0; i < row; ++i) {
            for (int j = 0; j < col; ++j) {
                if (result[i][j] > max_s) {
                    max_s = result[i][j];
                    max_i = i;
                    max_j = j;
                }
            }
        }

        System.out.printf("Max square size %d at %d:%d%n", max_s, max_i, max_j);
    }
}

File: hobbies/unsorted/max-square/max_square.py
#!/usr/bin/env python
from __future__ import print_function

def print_matrix(matrix):
    row, column = len(numbers), len(numbers[0])
    for i in xrange(row):
        for j in xrange(column):
            print(matrix[i][j], end = ' ')
        print('')

def max_square(numbers):
    print_matrix(numbers)
    row, column = len(numbers), len(numbers[0])
    result = [[0 for i in xrange(column)] for j in xrange(row)]
    for i in xrange(row):
        result[i][0] = numbers[i][0]

    for j in xrange(column):
        result[0][j] = numbers[0][j]

    for i in xrange(1, row):
        for j in xrange(1, column):
            if numbers[i][j] == 0:
                result[i][j] = 0
            else:
                result[i][j] = min((result[i][j-1], result[i-1][j], result[i-1][j-1])) + 1

    print('result')
    print_matrix(result)
    max_s, max_i, max_j = result[0][0], 0, 0
    for i in xrange(0, row):
        for j in xrange(0, column):
            if result[i][j] > max_s:
                max_s = result[i][j]
                max_i = i
                max_j = j
    print(i, j, max_s)


row_col = raw_input()
row, col = row_col.split()
numbers = []
for i in xrange(int(row)):
    line = raw_input()
    numbers.append([int(x) for x in line.split()])

max_square(numbers)
